src/PlusDataCollection/OpenHaptics/vtkPlusOpenHapticsDevice.cxx

```diff
@@ -0,0 +1,297 @@
+    {
+      LOG_WARNING("No Force tool in input channel.  Forces will not be available");
+    }
+  }
+
+  HDErrorInfo errorFlush;
+  while (HD_DEVICE_ERROR(errorFlush = hdGetError())) {}
+
+  DeviceHandle = hdInitDevice(this->DeviceName.c_str());
+
+  HDErrorInfo error;
+  if (HD_DEVICE_ERROR(error = hdGetError()))
+  {
+    LOG_ERROR( "Failed to initialize Phantom Omni " << this->DeviceName);
+    DeviceHandle = -1;
+    return PLUS_FAIL;
+  }
+
+  // Enable forces
+  hdEnable(HD_FORCE_OUTPUT);
+  hdEnable(HD_FORCE_RAMPING);
+
+  this->isHapticDeviceInitialized = true;
+  LOG_INFO("Phantom initialized: " << this->DeviceName)
+
+  hdStartScheduler();
+
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::InternalDisconnect()
+{
+  if (this->isHapticDeviceInitialized)
+  {
+    hdDisableDevice(DeviceHandle);
+    hdStopScheduler();
+    this->isHapticDeviceInitialized=false;
+  }
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::ReadConfiguration(vtkXMLDataElement* rootConfigElement)
+{
+  XML_FIND_DEVICE_ELEMENT_REQUIRED_FOR_READING(deviceConfig, rootConfigElement);
+  XML_READ_CSTRING_ATTRIBUTE_REQUIRED(DeviceName,deviceConfig);
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::WriteConfiguration(vtkXMLDataElement* rootConfigElement)
+{
+  XML_FIND_DEVICE_ELEMENT_REQUIRED_FOR_WRITING(trackerConfig, rootConfigElement);
+  trackerConfig->SetAttribute("DeviceName", this->DeviceName.c_str());
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::InternalStartRecording()
+{
+  if (!this->isHapticDeviceInitialized)
+  {
+    LOG_ERROR("InternalStartRecording failed: OpenHaptics device has not been initialized");
+    return PLUS_FAIL;
+  }
+  LOG_TRACE("vtkPlusOpenHapticsDevice::InternalStartRecording");
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::InternalStopRecording()
+{
+  LOG_TRACE("vtkPlusOpenHapticsDevice::InternalStopRecording");
+  return PLUS_SUCCESS;
+}
+
+HDCallbackCode HDCALLBACK
+vtkPlusOpenHapticsDevice::positionCallback(void* pData)
+{
+  
+  vtkPlusOpenHapticsDevice* client = reinterpret_cast<vtkPlusOpenHapticsDevice*>(pData);
+  HHD handle = client->DeviceHandle;
+  HD_state state = client->DeviceState;
+
+  const double unfilteredTimestamp = vtkPlusAccurateTimer::GetSystemTime();
+  ++client->FrameNumber;
+
+  HDdouble force[3];
+  force[0] = 0;
+  force[1] = 0;
+  force[2] = 0;
+
+  //Transform to RAS coordinate system
+  vtkSmartPointer<vtkMatrix4x4> rasCorrection = vtkSmartPointer<vtkMatrix4x4>::New();
+  rasCorrection->Identity();
+  rasCorrection->SetElement(0, 0, -1);
+  rasCorrection->SetElement(1, 1, 0);
+  rasCorrection->SetElement(1, 2, 1);
+  rasCorrection->SetElement(2, 2, 0);
+  rasCorrection->SetElement(2, 1, 1);
+
+  //assemble force data
+  auto inputChannels = client->InputChannels;
+  vtkPlusDataSource* forceInput;
+  if (!inputChannels.empty() && inputChannels[0]->GetOwnerDevice()->GetToolByPortName("Force", forceInput))

```

Comment:
Omit GetOwnerDevice, as you request a tool from a channel, not a device.

You are also guaranteed that InputChannels won't be empty, because you will have done that check in NotifiyConfigured
You can use client->InputChannels[0].GetTool(forceInput, "Force")

Also, be aware that most functions return PLUS_SUCCESS and PLUS_FAIL, so we generally explicitly compare to those values.

Reply:
I wanted to make it so the device could still run without a force input channel, so that's why I wanted to check the channels here.  Without the input channel I would just send a zero force vector to the device.

Reply:
Oh, I see. Ok, where would the force data come from?

Reply:
The use case for this is integrating the haptic device with 3D slicer.  The force (as the translation in a transform) comes from Slicer broadcasting as an OpenIGTLinkTracker device (I have config files for both the force and no force cases).

Reply:
Is there a method to get a tool based on portname without going through the device object?  The channels seem to only support getting a tool based on the transform ID.

Reply:
Accessor added! You can now get a tool by port name from a channel.

Reply:
awesome!